Mathlib/Analysis/NormedSpace/Star/ContinuousFunctionalCalculus/Instances.lean

```diff
@@ -245,43 +282,74 @@ lemma spectrum_star_mul_self_nonneg {b : A} : ∀ x ∈ spectrum ℝ (star b * b
     | one =>
       nontriviality A
       simp
-    | mul x _ y _ hx hy =>
+    | mul x x_mem y y_mem hx hy =>
       rw [← SpectrumRestricts.nnreal_iff] at hx hy ⊢
-      exact ⟨hx.1.add hy.1, hx.2.nnreal_add hx.1 hy.1 hy.2⟩
-  · rintro ⟨ha₁, ha₂⟩
-    let s := cfc Real.sqrt a
-    have : a = star s * s := by
-      rw [← cfc_id a (R := ℝ), ← cfc_star (R := ℝ) _ a, ← cfc_mul ..]
-      apply cfc_congr
-      peel ha₂ with x hx _
-      simp [Real.mul_self_sqrt this]
-    exact this ▸ star_mul_self_nonneg s
+      rw [← StarOrderedRing.nonneg_iff] at x_mem y_mem
+      exact hx.nnreal_add (.of_nonneg x_mem) (.of_nonneg y_mem) hy
 
-open NNReal in
-instance Nonneg.instContinuousFunctionalCalculus :
-    ContinuousFunctionalCalculus ℝ≥0 (fun x : A ↦ 0 ≤ x) :=
-  SpectrumRestricts.cfc (q := IsSelfAdjoint) ContinuousMap.realToNNReal
-    isometry_subtype_coe (fun _ ↦ nonneg_iff_isSelfAdjoint_and_spectrumRestricts)
-    (fun _ _ ↦ inferInstance)
+end NonnegSpectrumClass
 
-end Nonneg
+section RealEqComplex
+
+variable {A : Type*} [TopologicalSpace A] [Ring A] [StarRing A] [Algebra ℂ A]
+  [StarModule ℂ A] [ContinuousFunctionalCalculus ℂ (IsStarNormal : A → Prop)]
+  [ContinuousFunctionalCalculus ℝ (IsSelfAdjoint : A → Prop)]
+  [UniqueContinuousFunctionalCalculus ℝ A]
+  [UniqueContinuousFunctionalCalculus ℂ A]
+
+lemma cfcHom_real_eq_restrict {a : A} (ha : IsSelfAdjoint a) :
+    cfcHom ha = ha.spectrumRestricts.starAlgHom (cfcHom ha.isStarNormal) (f := Complex.reCLM) :=
+  have := UniqueContinuousFunctionalCalculus.compactSpace_spectrum a (R := ℂ)
+  ha.spectrumRestricts.cfcHom_eq_restrict Complex.isometry_ofReal ha ha.isStarNormal
+
+lemma cfc_real_eq_complex {a : A} (f : ℝ → ℝ) (ha : IsSelfAdjoint a := by cfc_tac)  :
+    cfc f a = cfc (fun x ↦ f x.re : ℂ → ℂ) a := by
+  have := UniqueContinuousFunctionalCalculus.compactSpace_spectrum a (R := ℂ)
+  replace ha : IsSelfAdjoint a := ha -- hack to avoid issues caused by autoParam
+  exact ha.spectrumRestricts.cfc_eq_restrict (f := Complex.reCLM) Complex.isometry_ofReal ha
+    ha.isStarNormal f
+
+end RealEqComplex
+
+section NNRealEqReal
+
+open NNReal
+
+variable {A : Type*} [TopologicalSpace A] [Ring A] [PartialOrder A] [StarRing A]
+  [StarOrderedRing A] [Algebra ℝ A] [TopologicalRing A] [ContinuousStar A] [StarModule ℝ A]
+  [ContinuousFunctionalCalculus ℝ (IsSelfAdjoint : A → Prop)]
+  [ContinuousFunctionalCalculus ℝ≥0 ((0 : A) ≤ ·)] [UniqueContinuousFunctionalCalculus ℝ A]
+  [NonnegSpectrumClass ℝ A]
+
+lemma cfcHom_nnreal_eq_restrict {a : A} (ha : 0 ≤ a) :
+    cfcHom ha = (SpectrumRestricts.nnreal_of_nonneg ha).starAlgHom
+      (cfcHom (IsSelfAdjoint.of_nonneg ha)) := by
+  have := UniqueContinuousFunctionalCalculus.compactSpace_spectrum a (R := ℝ)
+  apply (SpectrumRestricts.nnreal_of_nonneg ha).cfcHom_eq_restrict isometry_subtype_coe
+
+lemma cfc_nnreal_eq_real {a : A} (f : ℝ≥0 → ℝ≥0) (ha : 0 ≤ a := by cfc_tac)  :
+    cfc f a = cfc (fun x ↦ f x.toNNReal : ℝ → ℝ) a := by
```

Comment:
Do we also want `cfc_nnreal_eq_complex` for completeness?

Reply:
I think not? There is no direct coercion from NNReal to Complex anyway. As for the other direction, you could use nnnorm, but then you make it harder to stop midway in Real. If we use it a lot and it's a pain to go through both lemmas, then we can add a direct one.